Objects/genobject.c

```diff
@@ -58,7 +58,7 @@ gen_traverse(PyObject *self, visitproc visit, void *arg)
     PyGenObject *gen = _PyGen_CAST(self);
     Py_VISIT(gen->gi_name);
     Py_VISIT(gen->gi_qualname);
-    if (gen->gi_frame_state != FRAME_CLEARED) {
+    if (gen->gi_frame_state < FRAME_EXECUTING) {
```

Comment:
I *think* this is okay, but it also adds some additional constraints that will make the GC in the free threading build more fragile. I think we have to make sure that any gen/coro marked as `FRAME_EXECUTING` is on a `PyThreadState`'s frame stack -- otherwise some deferred references may not be visible to the GC and collected while still in use.

* There must not be any escaping calls between setting `gi_frame_state = FRAME_EXECUTING` and pushing the frame to the thread's stack.
* There must not be any escaping calls between popping the frame from the stack and setting `gi_frame_state` to some other value. (i.e., `exit_unwind` -> `clear_gen_frame`)

Reply:
At shutdown we delete all `PyThreadState`s except for the main thread. Those threads could be running generators or coroutines stuck in some long running call (like a `time.sleep()`). We run the GC after that. That seems like it would be unsafe (because we're now hiding deferred `_PyStackRef` from the GC.

Reply:
This test program segfaults in the free threading build with this PR:

https://gist.github.com/colesbury/11d59b9987e881a3c016b086bb4ba1ff

Reply:
If the generator is executing, it is part of the stack, not the heap.
So, I think `tp_traverse` should not be traversing executing generators.

@colesbury 
How does the free-threading GC find references that are on the stack (in normal frames)?
We should do that for executing generators.

Reply:
We loop over every `PyThreadState` in the interpreter and over the stack of frames in the thread state: https://github.com/python/cpython/blob/0cd21406bf84b3b4927a8117024232774823aee0/Python/gc_free_threading.c#L372-L373

During shutdown, we delete all `PyThreadState`s except for the main thread. So even though the generator is still marked as executing, it's not on any thread state's stack (because the `PyThreadState` [got deleted](https://github.com/python/cpython/blob/0cd21406bf84b3b4927a8117024232774823aee0/Python/pylifecycle.c#L2044-L2046) during `_Py_Finalize`)

Reply:
A generator that is marked as executing, but is not any thread's stack, is in an invalid state.

That sounds like the problem to me.
Do we know how it gets in that state?

Reply:
Yes, during shutdown, we delete all `PyThreadState`s except for the main thread.

Reply:
To answer my own question. We don't cleanup the references 🙁

Reply:
We call `PyThreadState_Clear()` on the deleted thread states, but that [doesn't clean up](https://github.com/python/cpython/blob/3d1df3d84e5c75a52b6f1379cd7f2809fc50befa/Python/pystate.c#L1639-L1651) `current_frame`. We don't call `PyThreadState_Delete()` for reasons that are not clear to me. Even if `PyThreadState_Clear()` cleaned up `current_frame` that wouldn't be sufficient because we unlink all the thread states before calling `PyThreadState_Clear()` and (by your definition) that already puts them in an invalid state. 

This is all longstanding CPython behavior, as far as I can tell. Changing the shutdown behavior seems likely to cause different shutdown-related bugs than what we experience today.

I think we can work around this by encoding more knowledge about generators in `gc_free_threading.c`. I don't love that, but it seems less risky than messing with the shutdown behavior. Let me know if you want to go that route -- I can help with the `gc_free_threading.c` changes.

Reply:
I'd like to investigate fixing up thread state clean up. While it may well risk introducing bugs in the short term, I think to would be better long term. It is hard to optimize anything if we can't trust our supposed invariants.

A quick fix, until we decide on how to handle this long term, would be to traverse the frame list when deleting threads and mark all generators as suspended.